cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java

```diff
@@ -17,53 +17,137 @@
+   * @param pageToken the token identifier for the current page requested in a paginated request
    * @param pageSize the number of application details returned in a paginated request
-   * @param filter optional filters in EBNF grammar. Currently Only one status and one type
-   *     filter is supported with AND expression.
+   * @param filter optional filters in EBNF grammar. Currently Only one status and one type filter
+   *     is supported with AND expression.
    */
   @GET
   @Path("/")
-  public void scanOperations(HttpRequest request, HttpResponder responder,
+  public void scanOperations(
+      HttpRequest request,
+      HttpResponder responder,
       @PathParam("namespace-id") String namespaceId,
       @QueryParam("pageToken") String pageToken,
       @QueryParam("pageSize") Integer pageSize,
-      @QueryParam("filter") String filter) {
-    // TODO(samik, CDAP-20812) fetch the operation runs from store
-    List<OperationRun> runs = new ArrayList<>();
-    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(runs));
+      @QueryParam("filter") String filter)
+      throws BadRequestException, IOException {
+    validateNamespace(namespaceId);
+    JsonPaginatedListResponder.respond(
+        GSON,
+        responder,
+        OPERATIONS_LIST_PAGINATED_KEY,
+        jsonListResponder -> {
+          AtomicReference<OperationRun> lastRun = new AtomicReference<>(null);
+          ScanOperationRunsRequest scanRequest =
+              getScanRequest(namespaceId, pageToken, pageSize, filter);
+          boolean pageLimitReached = false;
+          try {
+            pageLimitReached =
+                operationLifecycleManager.scanOperations(
+                    scanRequest,
+                    batchSize,
+                    runDetail -> {
+                      OperationRun run = runDetail.getRun();
+                      jsonListResponder.send(run);
+                      lastRun.set(run);
+                    });
+          } catch (IOException e) {
+            responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, e.getMessage());
+          } catch (OperationRunNotFoundException e) {
+            responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+          }
+          OperationRun run = lastRun.get();
+          return !pageLimitReached || run == null ? null : run.getId();
+        });
+  }
+
+  private ScanOperationRunsRequest getScanRequest(
+      String namespaceId, String pageToken, Integer pageSize, String filter) {
+    ScanOperationRunsRequest.Builder builder = ScanOperationRunsRequest.builder();
+    builder.setNamespace(namespaceId);
+    if (pageSize != null) {
+      builder.setLimit(pageSize);
+    }
+    if (pageToken != null) {
+      builder.setScanAfter(pageToken);
+    }
+    if (filter != null && !filter.isEmpty()) {
+      OperationType operationType = null;
+      OperationRunStatus operationStatus = null;
+      Pattern typePattern = Pattern.compile("type=(\\w+)");
```

Comment:
Can we please convert this to a function which parses pattern of key=val(ANDkey=val)* then returns the key,val pairs. Then we can match the key to type/status using switch case. if there is a unknown key we throw exception.